### PR TITLE
 Improve token pattern checking without validation 

### PR DESCRIPTION
**Pull request by adrianeboyd**
## Description

Add more detailed token pattern checks without full JSON pattern validation and
provide more detailed error messages.

Addresses #4070 (also related: #4063, #4100).

* Check whether top-level attributes in patterns and attr for PhraseMatcher are
  in token pattern schema

* Check whether attribute value types are supported in general (as opposed to
  per attribute with full validation)

* Report various internal error types (OverflowError, AttributeError, KeyError)
  as ValueError with standard error messages

* Check for tagger/parser in PhraseMatcher pipeline for attributes TAG, POS,
  LEMMA, and DEP

* Add error messages with relevant details on how to use validate=True or nlp()
  instead of nlp.make_doc()

* Support attr=TEXT for PhraseMatcher

* Add NORM to schema

* Expand tests for pattern validation, Matcher, PhraseMatcher, and EntityRuler

### Types of change

Some combination of bugfix and enhancement?

## Checklist

- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by adrianeboyd** [comment]
Concerns:

* `if attr not in TOKEN_PATTERN_SCHEMA["items"]["properties"].keys():` is potentially too hacky

* This will change the behavior for some users' existing faulty patterns that were silently discarding unsupported attributes

**Comment by ines** [comment]
> `if attr not in TOKEN_PATTERN_SCHEMA["items"]["properties"].keys()`: is potentially too hacky

Yeah, I see what you mean – but on the other hand, it only really refers to the standard JSON schmea structure, and a data structure that we control. It will not have to deal with any user input.

The `.keys()` should be redundant here btw so that'd at least make it slightly shorter.

> This will change the behavior for some users' existing faulty patterns that were silently discarding unsupported attributes

Yes, we shouldn't ship this in a patch release and only in v2.2. There are a few other PRs pending for the exact same reason.

**Comment by adrianeboyd** [comment]
Ah, I fixed the `.keys()` bits.

It seemed simpler to keep that check up-to-date with the structure than to extract a list, but it's still a bit hacky. I don't know how much the JSON schemas change across versions? But again, it's only 2 lines to keep in sync.

**Comment by ines** [comment]
> I don't know how much the JSON schemas change across versions? But again, it's only 2 lines to keep in sync.

I'm no JSON schemas expert, but I don't think the structure is going to change. And I think it's fine to use the schema as the "source of truth" for the matcher attributes – we just need to remember to update it if we ever change the supported attributes. 

(Sorry for rebuilding the PR btw but I'm trying to get this new WIP auto-merge-prevention bot to work and I thought I'd test it here. It's supposed to respect the "soon" emoji in the tag 😄)

**Comment by honnibal** [comment]
Looks good!

Only thing I noticed is that the wording on the errors seems a little different from the rest of the messages. I often avoid imperatives in the messages, but where we do use them, I think we usually don't use "please".

File: spacy/matcher/matcher.pyx [hunk 1]
```diff
@@ -568,6 +571,8 @@ def _preprocess_pattern(token_specs, string_store, extensions_table, extra_predi
             # Signifier for 'any token'
             tokens.append((ONE, [(NULL_ATTR, 0)], [], []))
             continue
+        if not isinstance(spec, dict):
+            raise ValueError(Errors.E154.format())
```
Comment (honnibal) [inline spacy/matcher/matcher.pyx]:
```suggestion
            raise ValueError(Errors.E158.format())
```

File: spacy/matcher/matcher.pyx [hunk 2]
```diff
@@ -581,21 +586,29 @@ def _get_attr_values(spec, string_store):
...
                 continue
-            elif attr.upper() == "OP":
+            elif attr == "OP":
                 continue
-            if attr.upper() == "TEXT":
+            if attr == "TEXT":
                 attr = "ORTH"
-            attr = IDS.get(attr.upper())
+            if attr not in TOKEN_PATTERN_SCHEMA["items"]["properties"]:
+                raise ValueError(Errors.E152.format(attr=attr))
```
Comment (honnibal) [inline spacy/matcher/matcher.pyx]:
```suggestion
                raise ValueError(Errors.E156.format(attr=attr))
```

File: spacy/matcher/matcher.pyx [hunk 3]
```diff
@@ -581,21 +586,29 @@ def _get_attr_values(spec, string_store):
...
+            attr = IDS.get(attr)
         if isinstance(value, basestring):
             value = string_store.add(value)
         elif isinstance(value, bool):
             value = int(value)
         elif isinstance(value, dict):
             continue
+        else:
+            raise ValueError(Errors.E153.format(vtype=type(value).__name__))
```
Comment (honnibal) [inline spacy/matcher/matcher.pyx]:
```suggestion
            raise ValueError(Errors.E157.format(vtype=type(value).__name__))
```

File: spacy/matcher/matcher.pyx [hunk 4]
```diff
@@ -581,21 +586,29 @@ def _get_attr_values(spec, string_store):
...
         if attr is not None:
             attr_values.append((attr, value))
+        else:
+            # should be caught above using TOKEN_PATTERN_SCHEMA
+            raise ValueError(Errors.E152.format(attr=attr))
```
Comment (honnibal) [inline spacy/matcher/matcher.pyx]:
```suggestion
            raise ValueError(Errors.E156.format(attr=attr))
```

File: spacy/matcher/matcher.pyx [hunk 5]
```diff
@@ -755,11 +768,13 @@ def _get_operators(spec):
...
     else:
         keys = ", ".join(lookup.keys())
-        raise KeyError(Errors.E011.format(op=spec["OP"], opts=keys))
+        raise ValueError(Errors.E011.format(op=spec["OP"], opts=keys))
 
 
 def _get_extensions(spec, string_store, name2index):
     attr_values = []
+    if not isinstance(spec.get("_", {}), dict):
+        raise ValueError(Errors.E154.format())
```
Comment (honnibal) [inline spacy/matcher/matcher.pyx]:
```suggestion
        raise ValueError(Errors.E158.format())
```

File: spacy/matcher/phrasematcher.pyx [hunk 6]
```diff
@@ -62,6 +63,11 @@ cdef class PhraseMatcher:
         if isinstance(attr, long):
             self.attr = attr
         else:
+            attr = attr.upper()
+            if attr == "TEXT":
+                attr = "ORTH"
+            if attr not in TOKEN_PATTERN_SCHEMA["items"]["properties"]:
+                raise ValueError(Errors.E152.format(attr=attr))
```
Comment (honnibal) [inline spacy/matcher/phrasematcher.pyx]:
```suggestion
                raise ValueError(Errors.E156.format(attr=attr))
```

File: spacy/matcher/phrasematcher.pyx [hunk 7]
```diff
@@ -123,6 +129,10 @@ cdef class PhraseMatcher:
             length = doc.length
             if length == 0:
                 continue
+            if self.attr in (POS, TAG, LEMMA) and not doc.is_tagged:
+                raise ValueError(Errors.E155.format())
```
Comment (honnibal) [inline spacy/matcher/phrasematcher.pyx]:
```suggestion
                raise ValueError(Errors.E159.format())
```

File: spacy/matcher/phrasematcher.pyx [hunk 8]
```diff
@@ -123,6 +129,10 @@ cdef class PhraseMatcher:
...
+            if self.attr == DEP and not doc.is_parsed:
+                raise ValueError(Errors.E156.format())
```
Comment (honnibal) [inline spacy/matcher/phrasematcher.pyx]:
```suggestion
                raise ValueError(Errors.E160.format())
```

**Pull request merged**